.github/workflows/build.yml

```diff
@@ -0,0 +1,271 @@
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*') }}
+          restore-keys: ${{ runner.os }}-gradle-
+      - name: Gradle build
+        uses: eskatos/gradle-command-action@v1
+        with:
+          arguments: --refresh-dependencies -PmodularBuild=${{ matrix.modular }} -PjunitVersion=${{ matrix.junit-version }} --stacktrace --scan clean build
+
+  # We want to be up to date and know issues with future java versions as soon as possible.
+  # Furthermore we also would love to see our build working with the latest gradle version.
+  # As those builds might fail, they are allowed to fail, and should not prevent merges.
+  experimental:
+    if: (contains(github.event.pull_request.labels.*.name, 'merge-ready') || !github.event.pull_request)
+    needs: full-featured
+    runs-on: ${{ matrix.os }}-latest
+    continue-on-error: true
+    timeout-minutes: 15
+    strategy:
+      matrix:
+        junit-version: [ '5.7.0', '5.7.1', '5.8.0-M1' ]
+        modular: [true, false]
+        os: [ubuntu, macos, windows]
+    name: Experimental build with java 17 and Gradle Release Candidate
+    steps:
+      - name: Check out repo
+        uses: actions/checkout@v2
+      - uses: sormuras/download-jdk@v1
+        with:
+          feature: ${{ env.EXPERIMENTAL_JAVA }}
+      - uses: actions/setup-java@v2
+        with:
+          java-version: ${{ env.JDK_VERSION }}
+          distribution: jdkfile
+          jdkFile: ${{ env.JDK_FILE }}
+      - name: Prepare JDK_EXPERIMENTAL env var
+        shell: bash
+        run: echo "JDK_EXPERIMENTAL=$JAVA_HOME" >> $GITHUB_ENV
+      - uses: actions/setup-java@v2
+        with:
+          java-version: 16
+          distribution: adopt
+      - name: Cache Gradle
+        uses: actions/cache@v2
+        with:
+          path: ~/.gradle/caches/
+          key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*') }}
+          restore-keys: ${{ runner.os }}-gradle-
+      - name: Gradle toolchains
+        uses: eskatos/gradle-command-action@v1
+        with:
+          gradle-version: rc
+          arguments: -Porg.gradle.java.installations.auto-download=false javaToolchains 
+      - name: Gradle build
+        uses: eskatos/gradle-command-action@v1
+        with:
+          gradle-version: rc
+          arguments: --refresh-dependencies -PexperimentalJavaVersion=17 -PmodularBuild=${{ matrix.modular }} -PjunitVersion=${{ matrix.junit-version }} -Porg.gradle.java.installations.auto-download=false --stacktrace --scan clean build
+
+  # Releasing will be only done, if there is a version defined and `releasing` is set to true.
+  # If not this stage will be ignored.
+  release:
+    if: github.event.inputs.version != '' && github.event.inputs.releasing == 'true'
+    runs-on: ubuntu-latest
+    needs: full-featured
+    timeout-minutes: 15
+    name: Release Pioneer into the wild
+    steps:
+      - name: Check out repo
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: '0'
```

Comment:
A comment would be helpful – just like in the case of Sonar – explaining why unshallow is needed (I also can't remember right now). Also, above we use an integer, here string.

Reply:
```suggestion
          # Sonar needs the whole git history for issue assignment
          fetch-depth: '0'
```

Reply:
I understand why we need it for the `sonarcloud` job, but why for the `release` job?

Reply:
for changelog generation